datafusion/jit/src/compile.rs

```diff
@@ -0,0 +1,184 @@
+pub fn build_calc_fn(
+    assembler: &Assembler,
+    jit_expr: JITExpr,
+    input_names: Vec<String>,
+) -> Result<GeneratedFunction> {
+    let mut builder = assembler.new_func_builder("calc_fn");
+    for input in &input_names {
+        builder = builder.param(format!("{}_array", input), I64);
+    }
+    let mut builder = builder.param("result", I64).param("len", I64);
+
+    let mut fn_body = builder.enter_block();
+    fn_body.declare_as("index", fn_body.lit_i(0))?;
+    fn_body.while_block(
+        |cond| cond.lt(cond.id("index")?, cond.id("len")?),
+        |w| {
+            w.declare_as("offset", w.mul(w.id("index")?, w.lit_i(PTR_SIZE as i64))?)?;
+            for input in &input_names {
+                w.declare_as(
+                    format!("{}_ptr", input),
+                    w.add(w.id(format!("{}_array", input))?, w.id("offset")?)?,
+                )?;
+                w.declare_as(input, w.deref(w.id(format!("{}_ptr", input))?, I64)?)?;
+            }
+            w.declare_as("res_ptr", w.add(w.id("result")?, w.id("offset")?)?)?;
+            w.declare_as("res", jit_expr.clone())?;
+            w.store(w.id("res")?, w.id("res_ptr")?)?;
+
+            w.assign("index", w.add(w.id("index")?, w.lit_i(1))?)?;
+            Ok(())
+        },
+    )?;
+
+    let gen_func = fn_body.build();
+    Ok(gen_func)
+}
+
+#[cfg(test)]
+mod test {
+    use std::{collections::HashMap, sync::Arc};
+
+    use arrow::{
+        array::{Array, PrimitiveArray},
+        datatypes::{DataType, Int64Type},
+    };
+    use datafusion_common::{DFField, DFSchema, DataFusionError};
+    use datafusion_expr::Expr as DFExpr;
+
+    use crate::ast::BinaryExpr;
+
+    use super::*;
+
+    fn run_df_expr(
+        df_expr: DFExpr,
+        schema: Arc<DFSchema>,
+        lhs: PrimitiveArray<Int64Type>,
+        rhs: PrimitiveArray<Int64Type>,
+    ) -> Result<PrimitiveArray<Int64Type>> {
+        if lhs.null_count() != 0 || rhs.null_count() != 0 {
+            return Err(DataFusionError::NotImplemented(
+                "Computing on nullable array not yet supported".to_string(),
+            ));
+        }
+        if lhs.len() != rhs.len() {
+            return Err(DataFusionError::NotImplemented(
+                "Computing on different length arrays not yet supported".to_string(),
+            ));
+        }
+
+        // translate DF Expr to JIT Expr
+        let input_fields = schema.field_names();
+        let jit_expr: JITExpr = (df_expr, schema).try_into()?;
+
+        // allocate memory for calc result
+        let len = lhs.len();
+        let result = vec![0i64; len];
+
+        // compile and run JIT code
+        let assembler = Assembler::default();
+        let gen_func = build_calc_fn(&assembler, jit_expr, input_fields)?;
+        let mut jit = assembler.create_jit();
+        let code_ptr = jit.compile(gen_func)?;
+        let code_fn =
+            unsafe { core::mem::transmute::<_, fn(i64, i64, i64, i64) -> ()>(code_ptr) };
```

Comment:
code like this is some of my favorite ❤️  -- cast some memory to a function pointer and call it ;)

Extremely unsafe but feels very powerful

Reply:
I wonder why not cast to the types you really want, like `fn(*i64, *i64, *i64, i64)` and then you can avoid the `as i64` stuff below

Reply:
Good suggestion 👍 

> Extremely unsafe but feels very powerful

Same!